### PR TITLE
Lowell 803 welcome page react

### PR DESCRIPTION
**Pull request by lowellrex**
Connects #803.

More work still to be done as part of #804 (which is blocked by #813 as well), but this PR just gets all the html elements in the correct spot looking somewhat passable visually.

### CODE REVIEW

File: client/package.json [hunk 1]
```diff
@@ -9,21 +9,30 @@
     "lint:fix": "yarn run lint --fix"
   },
   "dependencies": {
+    "@department-of-veterans-affairs/appeals-frontend-toolkit": "^2.1.0",
+    "babel-core": "^6.26.0",
```
Comment (nickheiner-usds) [inline client/package.json]: `babel-*` deps should really be on the `devDependencies` side, but it doesn't make a huge difference since this module will have no dependents.

File: client/src/components/AlertIcon.jsx [hunk 2]
```diff
@@ -0,0 +1,16 @@
+import React from 'react';
+
+export default class AlertIcon extends React.Component {
```
Comment (nickheiner-usds) [inline client/src/components/AlertIcon.jsx]: Please always default to `PureComponent`, since it avoids wasted renders.

File: client/src/components/RecentDownloadRow.jsx [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    // TODO: Use this.props.download.expiration_date or equivalent in UserManifest when it is built out.
+    const expirationDate = '01/26';
+
+    // TODO: Switch the link text based on this.props.download.in_progress or equivalent.
+    const downloadInProgress = true;
+
+    // TODO: Add alert icon based on this.props.download.had_errors or equivalent.
+    const downloadEncounteredErrors = true;
+
+    return <tr id={`download-${this.props.download.id}`}>
```
Comment (nickheiner-usds) [inline client/src/components/RecentDownloadRow.jsx]: Why do we need this `id`?

File: client/src/components/RecentDownloadRow.jsx [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+      <td>
+        {veteranFullName}
```
Comment (nickheiner-usds) [inline client/src/components/RecentDownloadRow.jsx]: You could also just do:

```jsx
{this.props.download.veteran_first_name} {this.props.download.veteran_last_name}
```

Whatever seems cleaner to you.

File: client/src/components/RecentDownloadRow.jsx [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        <span className="cf-subtext"> ({this.props.download.file_number}) - Expires on {expirationDate} </span>
+      </td>
+      <td className="ee-actions-cell">
+        <Link to={`/downloads/${this.props.download.id}`}>
+          { downloadEncounteredErrors ? <AlertIcon /> : '' }
```
Comment (nickheiner-usds) [inline client/src/components/RecentDownloadRow.jsx]: The idiomatic way to do this is:

```jsx
{ downloadEncounteredErrors && <AlertIcon /> }
```

File: client/src/containers/InitContainer.jsx [hunk 6]
```diff
@@ -1,28 +1,47 @@
...
+import WelcomeContainer from './WelcomeContainer';
 
+class InitContainer extends React.Component {
   render() {
-    return <div>
-      <textarea defaultValue={this.props.text} onChange={this.onTextareaChange.bind(this)} />
-      <h1>{this.props.text.length}</h1>
-    </div>;
+    return <BrowserRouter>
+      <div>
```
Comment (nickheiner-usds) [inline client/src/containers/InitContainer.jsx]: What is this `div` for?

File: client/src/containers/InitContainer.jsx [hunk 7]
```diff
@@ -1,28 +1,47 @@
...
+            <NavigationBar
+              appName="eFolder Express"
+              logoProps={{
+                accentColor: '#F0835e',
+                overlapColor: '#F0835e'
+              }}
+              userDisplayName={this.props.userDisplayName}
+              dropdownUrls={this.props.dropdownUrls}
+              defaultUrl="/react">
+              <div className="cf-wide-app section--document-list">
```
Comment (nickheiner-usds) [inline client/src/containers/InitContainer.jsx]: Do you really need `section--document-list`?

We must be sure that all class names and IDs are really needed. This is the easiest time to remove them if not; it's harder to do it in an error-free way once the code is older.

File: client/src/containers/WelcomeContainer.jsx [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+            </i></p>
+          </div>
+          <RecentDownloadsContainer />
+        </div>
+      </div>
+    </main>;
+  }
+}
+
+const mapStateToProps = (state) => {
```
Comment (nickheiner-usds) [inline client/src/containers/WelcomeContainer.jsx]: This can be simplified:

```js
const mapStateToProps = (state) => _.pick(state, 'authenticityToken');
```

Generally, please default to `() => expr` instead of `() => { return expr; }` when the function is just a single return value.

File: client/src/containers/RecentDownloadsContainer.jsx [hunk 9]
```diff
@@ -0,0 +1,35 @@
+import React from 'react';
+import { connect } from 'react-redux';
+
+import RecentDownloadRow from '../components/RecentDownloadRow';
+
+class RecentDownloadsContainer extends React.Component {
+  render() {
+    if (this.props.recentDownloads.length === 0) {
```
Comment (nickheiner-usds) [inline client/src/containers/RecentDownloadsContainer.jsx]:
```js
if (!this.props.recentDownloads.length) {
```

File: client/src/containers/WelcomeContainer.jsx (hunk 8 above)
Comment (nickheiner-usds) [reply client/src/containers/WelcomeContainer.jsx]: And please apply this change throughout the PR. :smile:

File: client/src/containers/WelcomeContainer.jsx [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+            <h1>Welcome to eFolder Express</h1>
+            <p>eFolder Express allows VA employees to bulk-download VBMS eFolders.
+              <br />Search for a Veteran ID number below to get started.
+            </p>
+          </div>
+          <div className="ee-search">
+            <form
+              className="usa-search usa-search-big cf-form"
+              id="new_download"
+              action="/downloads"
```
Comment (nickheiner-usds) [inline client/src/containers/WelcomeContainer.jsx]: This is not necessary. We're making an SPA now, so we're not going to use a `form action` which does a full page refresh. Instead, we'll make an XHR request to trigger the necessary server actions. That can be in a different PR. For now, let's just remove the unnecessary content.

File: client/src/containers/RecentDownloadsContainer.jsx [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+      <h2 className="ee-recent-searches">History</h2>
+      <table className="usa-table-borderless" summary="List of recent downloads and links to download their contents">
+        <thead>
+          <tr className="usa-sr-only">
+            <th scope="col">File Number</th>
+            <th scope="col">Actions</th>
+          </tr>
+        </thead>
+        <tbody>
+          {this.props.recentDownloads.map((dl, i) => <RecentDownloadRow download={dl} key={i} />)}
```
Comment (nickheiner-usds) [inline client/src/containers/RecentDownloadsContainer.jsx]: Please use longer variable names.

Using indexes for a key is bad, because it makes it hard for React to efficiently render the content. Are there any actual identifiers we can use?

**Comment by nickheiner-usds** [comment]
The page does not work:

image

(Ideally, the page should work before you submit it for CR. :smile:)

File: client/src/components/AlertIcon.jsx (hunk 2 above)
Comment (lowellrex) [reply client/src/components/AlertIcon.jsx]: For future reference, more information: https://stackoverflow.com/questions/41340697/react-component-vs-react-purecomponent#43936258

**Comment by nickheiner-usds** [comment]
I posted a message scolding you for breaking the page but it was my fault lol. Disregard. :smile:

File: client/src/components/RecentDownloadRow.jsx (hunk 3 above)
Comment (lowellrex) [reply client/src/components/RecentDownloadRow.jsx]: Grepping through the codebase `git grep download-` reveals that the reason we have this `id` in the table row element is so we can inspect it in tests. Going to leave the `id` around so we can maintain those tests.

File: client/src/components/RecentDownloadRow.jsx (hunk 4 above)
Comment (lowellrex) [reply client/src/components/RecentDownloadRow.jsx]: The v1 Download model [currently implements](https://github.com/department-of-veterans-affairs/caseflow-efolder/blob/b50f527fa8793edddb95d602d3fad7a80cf17aec/app/models/download.rb#L46) some (potentially incorrect?) logic when building `veteran_name`. I think it would make sense to replicate this logic in the new model on the backend (falling back to some sensible value other than `undef` if there is no last name) and just pass that forward to the frontend. Leaving this as a variable for now, to be replaced by the attribute when the backend implements that method.

File: client/src/containers/InitContainer.jsx (hunk 6 above)
Comment (lowellrex) [reply client/src/containers/InitContainer.jsx]: Good catch! It was left over from before `BrowserRouter` wrapped the child elements.

Comment (lowellrex) [reply client/src/containers/InitContainer.jsx]: I spoke too soon. We need it because `BrowserRouter` barfs with the following message if we don't have it, "A <Router> may have only one child element".

**Comment by nickheiner-usds** [comment]
Once my a-f-t (https://github.com/department-of-veterans-affairs/appeals-frontend-toolkit/pull/9) PR lands, here's a diff to use it.

```diff
diff --git a/client/src/containers/WelcomeContainer.jsx b/client/src/containers/WelcomeContainer.jsx
index a555474..0ea76ef 100644
--- a/client/src/containers/WelcomeContainer.jsx
+++ b/client/src/containers/WelcomeContainer.jsx
@@ -2,12 +2,13 @@ import React from 'react';
 import { connect } from 'react-redux';
 
 import RecentDownloadsContainer from './RecentDownloadsContainer';
+import AppSegment from '@department-of-veterans-affairs/appeals-frontend-toolkit/components/AppSegment';
 
 class WelcomeContainer extends React.PureComponent {
   render() {
     return <main className="usa-grid">
       <div className="cf-app">
-        <div className="cf-app-segment cf-app-segment--alt ee-new-download">
+        <AppSegment extraClassNames="ee-new-download" filledBackground>
           <div className="ee-heading">
             <h1>Welcome to eFolder Express</h1>
             <p>eFolder Express allows VA employees to bulk-download VBMS eFolders.
@@ -40,7 +41,7 @@ Note: eFolder Express now includes Virtual VA documents from the Legacy Content
             </i></p>
           </div>
           <RecentDownloadsContainer />
-        </div>
+        </AppSegment>
       </div>
     </main>;
   }

```

**Comment by nickheiner-usds** [comment]
The a-f-t PR landed, so please pull in `AppSegment` and use it here.

**Comment by nickheiner-usds** [comment]
There's a big margin at the bottom here that feels unnecessary. Let's look in person at how we can remove it:

image
